src/workerd/util/sentry.h

```diff
@@ -47,4 +48,42 @@ inline kj::StringPtr maybeOmitColoFromSentry(uint32_t coloId) {
     KJ_LOG(severity,  __VA_ARGS__); \
   }
 
-}  // namespace workerd
+// Log this to Sentry once ever per process. Typically will be better to use LOG_ERROR_PERIODICALLY.
+#define LOG_ERROR_ONCE(msg, ...)                                               \
+  do {                                                                         \
+    static bool logOnce KJ_UNUSED = [&]() {                                    \
+      KJ_LOG(ERROR, msg, ##__VA_ARGS__);                                       \
+      return true;                                                             \
+    }();                                                                       \
+  } while (0)
+
+#define LOG_ERROR_ONCE_IF(cond, msg, ...)                                      \
+  do {                                                                         \
+    if (cond) {                                                                \
+      LOG_ERROR_ONCE(msg, ##__VA_ARGS__);                                      \
+    }                                                                          \
+  } while (0)
+
+// Slightly more expensive than LOG_ERROR_ONCE. Avoid putting into a hot path (e.g. within a loop)
+// where an overhead of ~hundreds of nanoseconds per evaluation to retrieve the current time would
+// be prohibitive.
+#define LOG_ERROR_PERIODICALLY(msg, ...)                                       \
+  do {                                                                         \
+    static kj::TimePoint KJ_UNIQUE_NAME(lastLogged) =                          \
```

Comment:
In practice, this probably works, but this variable is static across all threads. You could either add the thread_local qualifier or place the TimePoint in an atomic.

Reply:
I noticed this too but didn't have a good solution. I like the thread_local idea. Cross-thread atomics feel grotty in terms of cross-cpu contention.

Reply:
Probably fine as long as it's acquire release and not CST? It's the memory fence that tends to really trip up threads. Still perfectly happy with thread local though.

Reply:
Is std::atomic<kj::TimePoint> going to be lock free? I feel like the STD library might inject a lock in there even if it's backed by a uint64_t under the hood... I guess could static assert that it's lock free to double-check.